trlx/orchestrator/ppo_orchestrator.py

```diff
@@ -105,52 +121,117 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):  # noq
+                    ref_logits[:, :-1, :], response_tensors[:, 1:]
+                )
+            else:
+                logprobs = logprobs_from_logits(logits[:, :-1, :], all_tokens[:, 1:])
+                ref_logprobs = logprobs_from_logits(
+                    ref_logits[:, :-1, :], all_tokens[:, 1:]
+                )
 
             n = samples.shape[0]
-            values = values.cpu()[:, :-1]
             logprobs = logprobs.cpu()
             ref_logprobs = ref_logprobs.cpu()
             query_tensors = query_tensors.cpu()
             response_tensors = response_tensors.cpu()
-
-            start = query_tensors.shape[1] - 1
-            ends = start + attention_mask[:, start:].sum(1)
-            all_values = [values[ix, start : ends[ix]] for ix in range(n)]
-            all_logprobs = [logprobs[ix, start : ends[ix]] for ix in range(n)]
+            if self.trainer.config.model.model_arch_type == "seq2seq":
+                start = 1  # skip the <s> token
+                ends = (response_tensors[:, start:] != 0).sum(1)
+                all_logprobs = [logprobs[ix, start : ends[ix]] for ix in range(n)]
+                all_values = [values[ix, start - 1 : ends[ix] - 1] for ix in range(n)]
+                rewards = [
+                    -self.trainer.kl_ctl.value
+                    * (
+                        logprobs[ix, start : ends[ix]]
+                        - ref_logprobs[ix, start : ends[ix]]
+                    )
+                    for ix in range(n)
+                ]
+            else:
+                n = samples.shape[0]
+                values = values.cpu()
+                logprobs = logprobs.cpu()
+                ref_logprobs = ref_logprobs.cpu()
+                query_tensors = query_tensors.cpu()
+                response_tensors = response_tensors.cpu()
+                start = (
+                    query_tensors.shape[1] - 1
+                )  # left shift by 1 ref: https://github.com/lvwerra/trl/blob/main/trl/trainer/ppo_trainer.py#L425
+                ends = start + attention_mask[:, start:].sum(1) - 1
+                for ix in range(n):
+                    if ends[ix] == all_tokens.shape[1]:
+                        ends[ix] = ends[ix] - 1
+                all_values = [values[ix, start - 1 : ends[ix] - 1] for ix in range(n)]
+                all_logprobs = [logprobs[ix, start : ends[ix]] for ix in range(n)]
+                rewards = -self.trainer.kl_ctl.value * (logprobs - ref_logprobs)
+                rewards = [rs[start : ends[ix]] for ix, rs in enumerate(rewards)]
 
             # Compute rewards
-            rewards = -self.trainer.kl_ctl.value * (logprobs - ref_logprobs)
             all_rewards = [None] * n
+
             for ix in range(n):
-                rs = rewards[ix][start : ends[ix]]
-                rs[-1] = scores[ix]
+                rs = rewards[ix]
+                if len(rs) == 0:
+                    rs = torch.tensor([0.0])
```

Comment:
Should we penalize empty responses? Also I can't understand how it's possible to have those unless `max_new_tokens == 0` 🤔

Reply:
Yes, this is a complete exception case, but I got it a few times when I ran PPO sentiments.

Reply:
Yeah Jon has said that he also experienced it. I wonder if the case of it is unknown it may be a symptom of some other bug elsewhere

Reply:
@reciprocated the only case I can think of that'd lead to an empty response is when `len(query)` is larger than the `generate` method's `min_length` arg, which defaults to `10`, and the model so happens to output the `eos_token` on its first sample. (Note that with causal models the `min_length` constraint includes the length of the context (query) meaning it won't actually have an effect on the generations if the min condition is already met by the context size).

Reply:
I do not think it is a serious issue. We can just throw an error if this min_length thing comes up. I've never seen this in practice when I set min length correctly. (Perhaps we should add an extra parameter called min_new_length...? We should upstream to HF transformers though)